File: ProgrammingProjects/cpp/Project_03/src/integrals.cpp
#include "minihf/integrals.h"

#include <algorithm>
#include <array>
#include <cmath>
#include <fstream>
#include <stdexcept>
#include <string>
#include <tuple>
#include <utility>

namespace minihf {
namespace {

struct MatrixEntry {
    Eigen::Index row;
    Eigen::Index column;
    double value;
};

std::ifstream open_input(const std::filesystem::path& path)
{
    std::ifstream input{path};
    if (!input) {
        throw std::runtime_error(
            "Cannot open required input file: " + path.string());
    }
    return input;
}

Eigen::MatrixXd read_symmetric_matrix(
    const std::filesystem::path& path,
    Eigen::Index expected_size = 0)
{
    std::ifstream input = open_input(path);
    std::vector<MatrixEntry> entries;
    Eigen::Index maximum_index = 0;

    long long file_row = 0;
    long long file_column = 0;
    double value = 0.0;

    while (input >> file_row >> file_column >> value) {
        if (file_row <= 0 || file_column <= 0) {
            throw std::runtime_error(
                "AO indices must start at 1 in " + path.string());
        }

        const Eigen::Index row =
            static_cast<Eigen::Index>(file_row - 1);
        const Eigen::Index column =
            static_cast<Eigen::Index>(file_column - 1);

        maximum_index = std::max(
            maximum_index,
            std::max(row, column) + 1);
        entries.push_back(MatrixEntry{row, column, value});
    }

    if (!input.eof()) {
        throw std::runtime_error(
            "Invalid matrix data in " + path.string());
    }
    if (entries.empty()) {
        throw std::runtime_error(
            "No matrix data found in " + path.string());
    }

    const Eigen::Index size =
        expected_size == 0 ? maximum_index : expected_size;
    if (maximum_index > size) {
        throw std::runtime_error(
            "AO index exceeds the expected basis size in " +
            path.string());
    }

    Eigen::MatrixXd matrix = Eigen::MatrixXd::Zero(size, size);
    for (const MatrixEntry& entry : entries) {
        matrix(entry.row, entry.column) = entry.value;
        matrix(entry.column, entry.row) = entry.value;
    }
    return matrix;
}

double read_scalar(const std::filesystem::path& path)
{
    std::ifstream input = open_input(path);
    double value = 0.0;
    if (!(input >> value)) {
        throw std::runtime_error(
            "Expected a number in " + path.string());
    }
    return value;
}

int read_electron_count(
    const std::filesystem::path& geometry_path,
    int molecular_charge)
{
    std::ifstream input = open_input(geometry_path);

    int atom_count = 0;
    if (!(input >> atom_count) || atom_count <= 0) {
        throw std::runtime_error(
            "Invalid atom count in " + geometry_path.string());
    }

    int total_nuclear_charge = 0;
    for (int atom = 0; atom < atom_count; ++atom) {
        double nuclear_charge = 0.0;
        double x = 0.0;
        double y = 0.0;
        double z = 0.0;

        if (!(input >> nuclear_charge >> x >> y >> z)) {
            throw std::runtime_error(
                "Not enough atom rows in " + geometry_path.string());
        }

        const long long rounded_charge = std::llround(nuclear_charge);
        if (std::abs(
                nuclear_charge - static_cast<double>(rounded_charge))
            > 1.0e-12) {
            throw std::runtime_error(
                "Non-integral nuclear charge in " +
                geometry_path.string());
        }
        total_nuclear_charge += static_cast<int>(rounded_charge);
    }

    return total_nuclear_charge - molecular_charge;
}

std::size_t dense_eri_size(Eigen::Index basis_size)
{
    const std::size_t size = static_cast<std::size_t>(basis_size);
    return size * size * size * size;
}

std::size_t flat_eri_index(
    Eigen::Index mu,
    Eigen::Index nu,
    Eigen::Index lambda,
    Eigen::Index sigma,
    Eigen::Index basis_size)
{
    const std::size_t size = static_cast<std::size_t>(basis_size);
    return (((static_cast<std::size_t>(mu) * size
              + static_cast<std::size_t>(nu))
             * size
             + static_cast<std::size_t>(lambda))
            * size
            + static_cast<std::size_t>(sigma));
}

std::vector<double> read_eri(
    const std::filesystem::path& path,
    Eigen::Index basis_size)
{
    std::ifstream input = open_input(path);
    std::vector<double> eri(dense_eri_size(basis_size), 0.0);

    long long file_i = 0;
    long long file_j = 0;
    long long file_k = 0;
    long long file_l = 0;
    double value = 0.0;

    while (input >> file_i >> file_j >> file_k >> file_l >> value) {
        const std::array<long long, 4> file_indices{
            file_i,
            file_j,
            file_k,
            file_l};
        for (const long long index : file_indices) {
            if (index <= 0 || index > basis_size) {
                throw std::runtime_error(
                    "ERI index is outside the AO basis in " +
                    path.string());
            }
        }

        const Eigen::Index i = static_cast<Eigen::Index>(file_i - 1);
        const Eigen::Index j = static_cast<Eigen::Index>(file_j - 1);
        const Eigen::Index k = static_cast<Eigen::Index>(file_k - 1);
        const Eigen::Index l = static_cast<Eigen::Index>(file_l - 1);

        const std::array<std::array<Eigen::Index, 4>, 8> permutations{{
            {{i, j, k, l}},
            {{j, i, k, l}},
            {{i, j, l, k}},
            {{j, i, l, k}},
            {{k, l, i, j}},
            {{l, k, i, j}},
            {{k, l, j, i}},
            {{l, k, j, i}},
        }};

        for (const auto& indices : permutations) {
            eri[flat_eri_index(
                indices[0],
                indices[1],
                indices[2],
                indices[3],
                basis_size)] = value;
        }
    }

    if (!input.eof()) {
        throw std::runtime_error(
            "Invalid ERI data in " + path.string());
    }
    return eri;
}

} // namespace

AOIntegrals::AOIntegrals(
    int electron_count,
    double nuclear_repulsion,
    Eigen::MatrixXd overlap,
    Eigen::MatrixXd kinetic,
    Eigen::MatrixXd nuclear_attraction,
    std::vector<double> electron_repulsion)
    : electron_count_{electron_count},
      nuclear_repulsion_{nuclear_repulsion},
      overlap_{std::move(overlap)},
      kinetic_{std::move(kinetic)},
      nuclear_attraction_{std::move(nuclear_attraction)},
      electron_repulsion_{std::move(electron_repulsion)}
{
    validate();
}

AOIntegrals AOIntegrals::from_directory(
    const std::filesystem::path& input_directory,
    int molecular_charge)
{
    if (!std::filesystem::is_directory(input_directory)) {
        throw std::invalid_argument(
            "Input directory does not exist: " +
            input_directory.string());
    }

    Eigen::MatrixXd overlap =
        read_symmetric_matrix(input_directory / "s.dat");
    const Eigen::Index basis_size = overlap.rows();

    return AOIntegrals{
        read_electron_count(
            input_directory / "geom.dat",
            molecular_charge),
        read_scalar(input_directory / "enuc.dat"),
        std::move(overlap),
        read_symmetric_matrix(
            input_directory / "t.dat",
            basis_size),
        read_symmetric_matrix(
            input_directory / "v.dat",
            basis_size),
        read_eri(input_directory / "eri.dat", basis_size)};
}

int AOIntegrals::electron_count() const noexcept
{
    return electron_count_;
}

Eigen::Index AOIntegrals::basis_size() const noexcept
{
    return overlap_.rows();
}

double AOIntegrals::nuclear_repulsion() const noexcept
{
    return nuclear_repulsion_;
}

const Eigen::MatrixXd& AOIntegrals::overlap() const noexcept
{
    return overlap_;
}

const Eigen::MatrixXd& AOIntegrals::kinetic() const noexcept
{
    return kinetic_;
}

const Eigen::MatrixXd& AOIntegrals::nuclear_attraction() const noexcept
{
    return nuclear_attraction_;
}

Eigen::MatrixXd AOIntegrals::core_hamiltonian() const
{
    return kinetic_ + nuclear_attraction_;
}

double AOIntegrals::eri(
    Eigen::Index mu,
    Eigen::Index nu,
    Eigen::Index lambda,
    Eigen::Index sigma) const
{
    return electron_repulsion_.at(
        eri_index(mu, nu, lambda, sigma));
}

std::size_t AOIntegrals::eri_index(
    Eigen::Index mu,
    Eigen::Index nu,
    Eigen::Index lambda,
    Eigen::Index sigma) const
{
    const Eigen::Index size = basis_size();
    for (const Eigen::Index index : {mu, nu, lambda, sigma}) {
        if (index < 0 || index >= size) {
            throw std::out_of_range("AO ERI index is outside the basis");
        }
    }
    return flat_eri_index(mu, nu, lambda, sigma, size);
}

void AOIntegrals::validate() const
{
    if (electron_count_ <= 0) {
        throw std::invalid_argument(
            "Electron count must be positive");
    }
    if (overlap_.rows() == 0 || overlap_.rows() != overlap_.cols()) {
        throw std::invalid_argument(
            "Overlap matrix must be non-empty and square");
    }

    const Eigen::Index size = overlap_.rows();
    const auto expected_shape = [size](const Eigen::MatrixXd& matrix) {
        return matrix.rows() == size && matrix.cols() == size;
    };

    if (!expected_shape(kinetic_)
        || !expected_shape(nuclear_attraction_)) {
        throw std::invalid_argument(
            "All one-electron matrices must have the same shape");
    }

    if (!overlap_.isApprox(overlap_.transpose())
        || !kinetic_.isApprox(kinetic_.transpose())
        || !nuclear_attraction_.isApprox(
            nuclear_attraction_.transpose())) {
        throw std::invalid_argument(
            "One-electron matrices must be symmetric");
    }

    if (electron_repulsion_.size() != dense_eri_size(size)) {
        throw std::invalid_argument(
            "Dense ERI storage has the wrong size");
    }
}

} // namespace minihf

File: ProgrammingProjects/python/chemistry/electronic_structure/integrals.py
"""AO integral data and readers for the Project 3 teaching inputs."""

from dataclasses import dataclass
from pathlib import Path

import numpy as np


@dataclass
class AOIntegrals:
    """All fixed input data needed by an electronic-structure method.

    The electron-repulsion integrals use chemist notation and dense
    storage::

        electron_repulsion[mu, nu, lam, sig]
            == (mu nu | lam sig)

    Dense storage is intentionally used in the Python teaching version.
    Project 3 is small enough that the clarity is worth more than ERI
    compression at this stage.
    """

    n_electrons: int
    nuclear_repulsion: float
    overlap: np.ndarray
    kinetic: np.ndarray
    nuclear_attraction: np.ndarray
    electron_repulsion: np.ndarray

    def __post_init__(self) -> None:
        self.overlap = np.asarray(self.overlap, dtype=float)
        self.kinetic = np.asarray(self.kinetic, dtype=float)
        self.nuclear_attraction = np.asarray(
            self.nuclear_attraction,
            dtype=float,
        )
        self.electron_repulsion = np.asarray(
            self.electron_repulsion,
            dtype=float,
        )
        self._validate()

    @property
    def n_basis(self) -> int:
        return self.overlap.shape[0]

    @property
    def core_hamiltonian(self) -> np.ndarray:
        """Return the one-electron core Hamiltonian H_core = T + V."""
        return self.kinetic + self.nuclear_attraction

    def _validate(self) -> None:
        if self.n_electrons <= 0:
            raise ValueError("n_electrons must be positive")

        if self.overlap.ndim != 2:
            raise ValueError("overlap must be a matrix")

        n_basis, n_columns = self.overlap.shape
        if n_basis != n_columns:
            raise ValueError("overlap must be square")

        matrix_shape = (n_basis, n_basis)
        for name, matrix in (
            ("kinetic", self.kinetic),
            ("nuclear_attraction", self.nuclear_attraction),
        ):
            if matrix.shape != matrix_shape:
                raise ValueError(
                    f"{name} has shape {matrix.shape}; "
                    f"expected {matrix_shape}"
                )

        eri_shape = (n_basis, n_basis, n_basis, n_basis)
        if self.electron_repulsion.shape != eri_shape:
            raise ValueError(
                "electron_repulsion has shape "
                f"{self.electron_repulsion.shape}; "
                f"expected {eri_shape}"
            )

        for name, matrix in (
            ("overlap", self.overlap),
            ("kinetic", self.kinetic),
            ("nuclear_attraction", self.nuclear_attraction),
        ):
            if not np.allclose(matrix, matrix.T):
                raise ValueError(f"{name} must be symmetric")


def load_ao_integrals(
    input_directory: str | Path,
    molecular_charge: int = 0,
) -> AOIntegrals:
    """Load the Project 3 integral files from one input directory."""
    directory = Path(input_directory)
    if not directory.is_dir():
        raise ValueError(f"Input directory does not exist: {directory}")

    overlap = _read_symmetric_matrix(directory / "s.dat")
    n_basis = overlap.shape[0]

    kinetic = _read_symmetric_matrix(
        directory / "t.dat",
        expected_size=n_basis,
    )
    nuclear_attraction = _read_symmetric_matrix(
        directory / "v.dat",
        expected_size=n_basis,
    )
    electron_repulsion = _read_eri(
        directory / "eri.dat",
        n_basis=n_basis,
    )
    nuclear_repulsion = _read_scalar(directory / "enuc.dat")
    n_electrons = (
        _read_total_nuclear_charge(directory / "geom.dat")
        - molecular_charge
    )

    return AOIntegrals(
        n_electrons=n_electrons,
        nuclear_repulsion=nuclear_repulsion,
        overlap=overlap,
        kinetic=kinetic,
        nuclear_attraction=nuclear_attraction,
        electron_repulsion=electron_repulsion,
    )


def _read_symmetric_matrix(
    path: Path,
    expected_size: int | None = None,
) -> np.ndarray:
    rows = _read_numeric_rows(path, expected_columns=3)
    inferred_size = max(max(int(row[0]), int(row[1])) for row in rows)

    if expected_size is not None and inferred_size > expected_size:
        raise ValueError(
            f"Index in {path} exceeds basis size {expected_size}"
        )

    size = expected_size or inferred_size
    matrix = np.zeros((size, size), dtype=float)

    for raw_i, raw_j, value in rows:
        i = int(raw_i) - 1
        j = int(raw_j) - 1
        _check_index(i, size, path)
        _check_index(j, size, path)
        matrix[i, j] = value
        matrix[j, i] = value

    return matrix


def _read_eri(path: Path, n_basis: int) -> np.ndarray:
    rows = _read_numeric_rows(path, expected_columns=5)
    eri = np.zeros((n_basis,) * 4, dtype=float)

    for raw_i, raw_j, raw_k, raw_l, value in rows:
        i, j, k, l = (
            int(raw_i) - 1,
            int(raw_j) - 1,
            int(raw_k) - 1,
            int(raw_l) - 1,
        )
        for index in (i, j, k, l):
            _check_index(index, n_basis, path)

        permutations = (
            (i, j, k, l),
            (j, i, k, l),
            (i, j, l, k),
            (j, i, l, k),
            (k, l, i, j),
            (l, k, i, j),
            (k, l, j, i),
            (l, k, j, i),
        )
        for permutation in permutations:
            eri[permutation] = value

    return eri


def _read_scalar(path: Path) -> float:
    text = _require_file(path).read_text(encoding="utf-8").strip()
    try:
        return float(text)
    except ValueError as error:
        raise ValueError(f"Expected one number in {path}") from error


def _read_total_nuclear_charge(path: Path) -> int:
    lines = _require_file(path).read_text(encoding="utf-8").splitlines()
    if not lines:
        raise ValueError(f"Geometry file is empty: {path}")

    try:
        n_atoms = int(lines[0].strip())
    except ValueError as error:
        raise ValueError(
            f"First line of {path} must be the atom count"
        ) from error

    atom_lines = [line for line in lines[1:] if line.strip()]
    if len(atom_lines) != n_atoms:
        raise ValueError(
            f"{path} declares {n_atoms} atoms but contains "
            f"{len(atom_lines)} atom rows"
        )

    total_charge = 0
    for line_number, line in enumerate(atom_lines, start=2):
        fields = line.split()
        if not fields:
            continue
        try:
            nuclear_charge = float(fields[0])
        except ValueError as error:
            raise ValueError(
                f"Invalid nuclear charge in {path}:{line_number}"
            ) from error

        rounded_charge = round(nuclear_charge)
        if not np.isclose(nuclear_charge, rounded_charge):
            raise ValueError(
                f"Non-integral nuclear charge in {path}:{line_number}"
            )
        total_charge += int(rounded_charge)

    return total_charge


def _read_numeric_rows(
    path: Path,
    expected_columns: int,
) -> list[tuple[float, ...]]:
    rows: list[tuple[float, ...]] = []
    for line_number, line in enumerate(
        _require_file(path).read_text(encoding="utf-8").splitlines(),
        start=1,
    ):
        if not line.strip():
            continue

        fields = line.split()
        if len(fields) != expected_columns:
            raise ValueError(
                f"Expected {expected_columns} columns in "
                f"{path}:{line_number}; got {len(fields)}"
            )
        try:
            rows.append(tuple(float(field) for field in fields))
        except ValueError as error:
            raise ValueError(
                f"Invalid number in {path}:{line_number}"
            ) from error

    if not rows:
        raise ValueError(f"No data found in {path}")
    return rows


def _require_file(path: Path) -> Path:
    if not path.is_file():
        raise ValueError(f"Required input file does not exist: {path}")
    return path


def _check_index(index: int, size: int, path: Path) -> None:
    if not 0 <= index < size:
        raise ValueError(
            f"AO index {index + 1} in {path} is outside 1..{size}"
        )
